# 19.自定义异常

1. 为什么需要自定义异常？
  1. 部分属性诸如性别年龄,在API中没有定义错误,需要根据我们程序需要定义我们需要的自定义异常
2. 自定义异常的步骤是什么？
  1. 继承Exception或RuntimeException
  1. 定义构造方法
  1. 使用异常

![image.png](https://cdn.nlark.com/yuque/0/2019/png/349894/1559703327098-da622b26-4648-4e54-aa26-793ac7fc30a8.png#align=left&display=inline&height=85&name=image.png&originHeight=85&originWidth=399&size=36803&status=done&width=399)<br />如性别,年龄,api中没有定义,我们要定义我们所需的

![image.png](https://cdn.nlark.com/yuque/0/2019/png/349894/1559703378706-8914dbdc-b2ff-47a7-880d-0162ceffdab9.png#align=left&display=inline&height=118&name=image.png&originHeight=118&originWidth=313&size=25751&status=done&width=313)<br />1.继承Exception或RuntimeException<br />2.定义构造方法<br />3.使用异常

1.继承<br />![image.png](https://cdn.nlark.com/yuque/0/2019/png/349894/1559703426933-e87064ff-1e3f-4d8e-b99c-c5fb4c38373b.png#align=left&display=inline&height=63&name=image.png&originHeight=63&originWidth=418&size=20880&status=done&width=418)<br />2.编写构造方法<br />![image.png](https://cdn.nlark.com/yuque/0/2019/png/349894/1559703459425-e15b46d9-2cd0-4afc-963f-c13e04f9c349.png#align=left&display=inline&height=356&name=image.png&originHeight=356&originWidth=275&size=100858&status=done&width=275)<br />![image.png](https://cdn.nlark.com/yuque/0/2019/png/349894/1559703488074-4e704e5e-5180-4b5f-81d6-0d875970f2a7.png#align=left&display=inline&height=247&name=image.png&originHeight=247&originWidth=372&size=47350&status=done&width=372)<br />定义完成<br />![image.png](https://cdn.nlark.com/yuque/0/2019/png/349894/1559703520168-202975df-bf21-487a-8f9a-d2a1968d6f40.png#align=left&display=inline&height=208&name=image.png&originHeight=208&originWidth=448&size=79069&status=done&width=448)<br />编写测试类<br />![image.png](https://cdn.nlark.com/yuque/0/2019/png/349894/1559703623970-0a286592-f198-455a-a556-797c30eeb05f.png#align=left&display=inline&height=177&name=image.png&originHeight=177&originWidth=456&size=106849&status=done&width=456)<br />不符合要求抛出异常,可以在方法出声明<br />![image.png](https://cdn.nlark.com/yuque/0/2019/png/349894/1559703671695-da4ad891-9524-41bd-815c-b4d6fca4fc08.png#align=left&display=inline&height=159&name=image.png&originHeight=159&originWidth=502&size=92320&status=done&width=502)<br />或者使用try catch<br />![image.png](https://cdn.nlark.com/yuque/0/2019/png/349894/1559703756384-75271a55-bf90-4968-9a4d-01cb54678835.png#align=left&display=inline&height=146&name=image.png&originHeight=146&originWidth=401&size=68396&status=done&width=401)<br />执行<br />![image.png](https://cdn.nlark.com/yuque/0/2019/png/349894/1559703815936-6404aa29-31ef-447b-a5f6-feb6219076a7.png#align=left&display=inline&height=131&name=image.png&originHeight=131&originWidth=649&size=57455&status=done&width=649)<br />最后的是标准错误流进行打印输出的描述
